pkg/controller/macvlannetwork/macvlannetwork_controller.go

```diff
@@ -0,0 +1,248 @@
+	kscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	netattdefv1 "github.com/Mellanox/network-operator/pkg/apis/k8s/v1"
+	mellanoxv1alpha1 "github.com/Mellanox/network-operator/pkg/apis/mellanox/v1alpha1"
+)
+
+const (
+	LASTNETWORKNAMESPACE = "operator.macvlannetwork.mellanx.io/last-network-namespace"
+	FINALIZERNAME        = "netattdef.finalizers.macvlannetwork.mellanox.com"
+)
+
+var log = logf.Log.WithName("controller_macvlannetwork")
+
+// Add creates a new MacvlanNetwork Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileMacvlanNetwork{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("macvlannetwork-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource MacvlanNetwork
+	err = c.Watch(&source.Kind{Type: &mellanoxv1alpha1.MacvlanNetwork{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource NetworkAttachmentDefinition
+	err = c.Watch(&source.Kind{Type: &netattdefv1.NetworkAttachmentDefinition{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileMacvlanNetwork implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileMacvlanNetwork{}
+
+// ReconcileMacvlanNetwork reconciles a MacvlanNetwork object
+type ReconcileMacvlanNetwork struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a MacvlanNetwork object and makes changes based on the state read
+// and what is in the MacvlanNetwork.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileMacvlanNetwork) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
```

Comment:
can we do the same "state" pattern as in nicclusterpolicy controller ?
i.e handle reconcile of an object in its own state under pkg

WDYT?

Reply:
Change to use "state" pattern

Reply:
as macvlan network is a cluster resource, it is now possible to set to net-attach-def controller reference for cleanup.